### ISSUE TITLE
using match on Maybe<ParentType>

### ISSUE THREAD

**Issue by ekiwi** [StanzaOrg/lbstanza-old#41]
I am new to Stanza and sometimes just try out things that seem intuitive.
The following code compiles, but does not behave the way that I would expect it to behave:

```
deftype Animal
defstruct Dog <: Animal
defstruct Cat <: Animal

defn check-for-animal (a:Maybe<Animal>):
    match(a):
        (d:One<Dog>) : println("it's a dog")
        (c:One<Cat>) : println("it's a cat")
        (n:None)     : println("nothing to see")

check-for-animal(One(Cat()))
check-for-animal(One(Dog()))
check-for-animal(None())
```

this prints

```
it's a dog
it's a dog
nothing to see
```

When I try to determine the type of `One(Cat())` (e.g. by compiling `var x:Int = One(Cat())`), the compiler confirms that the type is `One<Cat>`. So why is it cast to an object of type `One<Dog>` by `match`?
What am I missing here?

**Comment by CuppoJava** [comment]
Stanza's match construct (and method dispatch mechanisms) can only dispatch based on the top-level type. So in your code:

```
match(a):
    (d:One<Dog>) : println("it's a dog")
    (c:One<Cat>) : println("it's a cat")
    (n:None)     : println("nothing to see")
```

The first branch is always taken if the object is a One object, regardless of whether it internally contains a Dog or Cat object. This behaviour is by design. In the future, the compiler will issue a warning for attempting to dispatch based on the parametric properties of a type.

  -Patrick

**Comment by ekiwi** [comment]
Hi Patrick,

thank you very much for the explanation.

> based on the top-level type
> The first branch is always taken if the object is a One object,

I come from a C++ background where a template class with a different argument creates a new type. So this is one of the reasons why I thought that this might work.

> In the future, the compiler will issue a warning for attempting to dispatch based on the parametric properties of a type.

Why not an error?
To me the situation is as follows: I tell the compiler that I want to have an object of the type `One<Dog>` but the only thing that it guarantee me is that I get an object of the top level type `One`.

So basically, if I understand you correctly, the following is equivalent to the match statement above:

```
    match(a):
        (d:One)  : println("it's a dog")
        (d:One)  : println("it's a cat")     ; this should probably cause a compiler warning
        (n:None) : println("nothing to see")
```

So personally I would prefer if the compiler just requires the pattern to be a top level type. This way match actually guarantees me that I get the type that I wrote down and not the type that the compiler simplifies my pattern to.

**Comment by CuppoJava** [comment]
Hi Kevin.

Yes you're right. We should just make it an error instead of a warning.

If you're curious, here's a little more explanation as to why we designed it this way. Dispatching based on only the top-level type is necessary for ensuring that users are able to freely transition between untyped and typed code. For example, let's say that the user creates an array for holding integers

```
val xs = Array<Int>(10)
```

, and gets the program working. Then later, (for prototyping purposes,) he needs to temporarily store some strings in the array, so he goes and indicates that `xs` can hold anything.

```
val xs = Array<?>(10)
```

Currently, Stanza guarantees that you can make this change without affecting the behaviour of the program. But if we allowed you to dispatch based upon the parametric properties of a type then this change could lead to you dispatching to a different chunk of code, thus altering the behaviour of the program. 

  -Patrick
